### PR TITLE
Fix push_to_hub by not calling create_branch if branch exists

### LINKED ISSUE #7067: Convert_to_parquet fails for datasets with multiple configs
If the dataset has multiple configs, when using the `datasets-cli convert_to_parquet` command to avoid issues with the data viewer caused by loading scripts, the conversion process only successfully converts the data corresponding to the first config. When it starts converting the second config, it throws an error:

```
Traceback (most recent call last):
  File "/opt/anaconda3/envs/dl/bin/datasets-cli", line 8, in <module>
    sys.exit(main())
  File "/opt/anaconda3/envs/dl/lib/python3.10/site-packages/datasets/commands/datasets_cli.py", line 41, in main
    service.run()
  File "/opt/anaconda3/envs/dl/lib/python3.10/site-packages/datasets/commands/convert_to_parquet.py", line 83, in run
    dataset.push_to_hub(
  File "/opt/anaconda3/envs/dl/lib/python3.10/site-packages/datasets/dataset_dict.py", line 1713, in push_to_hub
    api.create_branch(repo_id, branch=revision, token=token, repo_type="dataset", exist_ok=True)
  File "/opt/anaconda3/envs/dl/lib/python3.10/site-packages/huggingface_hub/utils/_validators.py", line 114, in _inner_fn
    return fn(*args, **kwargs)
  File "/opt/anaconda3/envs/dl/lib/python3.10/site-packages/huggingface_hub/hf_api.py", line 5503, in create_branch
    hf_raise_for_status(response)
  File "/opt/anaconda3/envs/dl/lib/python3.10/site-packages/huggingface_hub/utils/_errors.py", line 358, in hf_raise_for_status
    raise BadRequestError(message, response=response) from e
huggingface_hub.utils._errors.BadRequestError:  (Request ID: Root=1-669fc665-7c2e80d75f4337496ee95402;731fcdc7-0950-4eec-99cf-ce047b8d003f)

Bad request:
Invalid reference for a branch: refs/pr/1
```

### PR DESCRIPTION
**Pull request by albertvillanova**
Fix push_to_hub by not calling create_branch if branch exists.

Note that currently create_branch raises a 403 Forbidden error even if all these conditions are met:
- exist_ok is passed
- the branch already exists
- the user does not have WRITE permission

Fix #7067.

Related issue:
- https://github.com/huggingface/huggingface_hub/issues/2419

### CODE REVIEW

**Comment by lhoestq** [comment]
cc @Wauplin maybe it's a `huggingface_hub` bug ?

File: src/datasets/arrow_dataset.py [hunk 1]
```diff
@@ -5620,7 +5620,8 @@ def push_to_hub(
         )
         repo_id = repo_url.repo_id
 
-        if revision is not None:
+        if revision is not None and not api.revision_exists(repo_id, revision, repo_type="dataset", token=token):
```
Comment (albertvillanova) [inline src/datasets/arrow_dataset.py:5623]: Alternatively:
```suggestion
        if revision is not None and not revision.startswith("refs/pr/"):
```

**Comment by albertvillanova** [comment]
I think we need to make this fix anyway, unless we pin the lower version of huggingface-hub (once they release the patch).

File: src/datasets/arrow_dataset.py (hunk 1 above)
Comment (albertvillanova) [reply src/datasets/arrow_dataset.py:5623]: I think this modification would be more efficient:
- The modification above will fix the 400 Bad Request error without the additional network call `revision_exists`.
- In the case of a 403 Forbidden error, anyway we will raise some code lines below when trying to push to the branch...

Comment (albertvillanova) [reply src/datasets/arrow_dataset.py:5623]: On the other hand, it is true that the call to `revision_exists` is a more general solution.

**Comment by albertvillanova** [comment]
Comment by @Wauplin: https://github.com/huggingface/huggingface_hub/pull/2426#issuecomment-2257657543
> I think this should be something to fix in datasets directly. Having a 400 Bad request when trying to create the branch refs/pr/1 seems normal to me since it's not a branch.

**Comment by lhoestq** [comment]
does this mean we should use `create_pull_request()` in that case ?

**Comment by Wauplin** [comment]
> does this mean we should use create_pull_request() in that case ?

If user wants to push some data to a new PR, they can already pass `create_pr=True` which will automatically do the job for you (without using `revision`). If user is passing `revision="refs/pr/1"` explicitly, you should assume the PR already exists.

**Comment by lhoestq** [comment]
ah yes we do pass create_pr in `preupload_lfs_files()` ! sounds good then

**Review by lhoestq — APPROVED** [review]

**Pull request merged**